wpengine/faustjs
# Show warning for active incompatible plugins

## Description



Adds a warning notice for users when an active plugin is known to conflict with FaustWP.

- The warning is shown on the plugins page and on the FaustWP settings page.
- Any user with the `activate_plugins` capability can see the warning and dismiss it. Dismissal occurs on a per user basis, meaning one user's dismissal does not remove the warning for other users.
- Only the conflicts that are present in the warning message at the time of dismissal are saved. If new conflicts are added to the list or a previously inactive conflict becomes active, the notice will be shown again to report the new conflicts.
- If a user deactivates FaustWP, their dismissed conflicts are cleared.

### What is not covered:

- Multisite. Accounting for network activations, network screens, super admin capabilities, and network notice dismissals added significant complexity.
- Must-use plugins.
- Themes. The ticket AC included themes. Seeing as we don't have any incompatible themes to report at the moment, it seemed unnecessary to create the alternate logic. I also question if any theme could be considered truly compatible with a decoupled architecture 🤔?

## Testing



I've added integration tests for most of the core functionality. However, the callbacks to display and dismiss the notice aren't easily testable in our current setup. I'd recommend testing with the following procedure.

With FaustWP installed and active:

1. Install and activate Elementor from Plugins -> Add New
2. Go to Settings -> Headless
3. See that the warning notice is present.
4. Go to Plugins -> Installed Plugins
5. See that the warning notice is present.
6. Deactivate Elementor
7. See the notice is no longer present.
8. Activate Elementor
9. See the notice returns
10. Dismiss the notice and refresh the page
11. See that the notice is gone
12. Deactivate and re-activate FaustWP.
13. See that the notice returns
14. Dismiss the notice
15. Try adding `wp-graphql/wp-graphql.php` to [the conflict list](https://github.com/wpengine/faustjs/blob/c9839d6b01d2ec5ed2bc3b7380304342aec78d84/plugins/faustwp/includes/detect-conflicts/functions.php#L21-L23)
16. Make sure WP GraphQL is active.
17. Refresh the plugins list page.
18. See that the notice returns and WP GraphQL is the only plugin listed.
19. Dismiss the notice and refresh the page
20. See that the notice is gone

## Screenshots


<img width="750" alt="Screen Shot 2022-04-22 at 3 19 30 PM" src="https://user-images.githubusercontent.com/4661832/164783746-af3309ae-59b8-4c3c-b7a0-4f6d186b40b8.png">

## Documentation Changes



I've created a follow up ticket (MERL-282) to add an "Incompatible plugins" page to the documentation. We should link users to that page from this warning notice to provide context and suggest workarounds when available.
